pilot/pkg/xds/ads.go

```diff
@@ -240,16 +243,30 @@ func (s *DiscoveryServer) Stream(stream DiscoveryStream) error {
 
 // update the node associated with the connection, after receiving a packet from envoy, also adds the connection
 // to the tracking map.
-func (s *DiscoveryServer) initConnection(node *core.Node, con *Connection, identities []string) error {
+func (s *DiscoveryServer) initConnection(node *core.Node, con *Connection, identities *security.Caller) error {
 	// Setup the initial proxy metadata
 	proxy, err := s.initProxyMetadata(node)
 	if err != nil {
 		return err
 	}
+
 	// Check if proxy cluster has an alias configured, if yes use that as cluster ID for this proxy.
+	origClusterID := proxy.Metadata.ClusterID
 	if alias, exists := s.ClusterAliases[proxy.Metadata.ClusterID]; exists {
 		proxy.Metadata.ClusterID = alias
 	}
+	if features.CentralIstiodAccess {
+		if identities.ClusterID != "" {
+			if string(proxy.Metadata.ClusterID) != identities.ClusterID && string(origClusterID) != identities.ClusterID {
+				return errors.New("cluster ID in node and auth not matching")
+			}
+		} else {
+			// This may become a hard error - it may allow exposing secrets from other clusters without verification.
+			// For backward compat and initially just a log.
+			log.WithLabels("method", identities.AuthSource).Info("Can't validate cluster ID")
+		}
+	}
+
```

Comment:
nit, we probably don't need the else line and we should consider adding the two IDs and add a debug log. Wdyt? Something like the following (I'm not sure of the order here and the format could be better but I think it still gets the point across). It seems like if IDs don't match then we might want to surface that information via a debug format log for easier debugging:

```suggestion
	if features.CentralIstiodAccess {
		if identities.ClusterID != "" {
			if string(proxy.Metadata.ClusterID) != identities.ClusterID && string(origClusterID) != identities.ClusterID {
				log.Debugf("cluster ID in node [%v] and cluster ID in auth [%v], do not match", identities.ClusterID, string(origClusterID))
				return errors.New("cluster ID in node and auth not matching")
			}
		}
		// This may become a hard error - it may allow exposing secrets from other clusters without verification.
		// For backward compat and initially just a log.
		log.WithLabels("method", identities.AuthSource).Info("Can't validate cluster ID")
	}

```